support 'west init --topdir'

thorsten-klein opened the pull request:
Fixes #774  (Successor of #775)

This PR introduces a new `--topdir` option for `west init`, allowing users to explicitly specify the west workspace directory (topdir) during initialization.  

This is particularly useful when using `--local`, as the current behavior always sets the parent of the given directory as the topdir. In addition, west is currently unable to determine both `manifest.path` and `manifest.file` from a single `directory` argument as the path can be interpreted in multiple ways (e.g. it can be split at various places).

With the new `--topdir` argument, users can now directly control where the west workspace is created.  

Parameterized tests for `west init --topdir` are added as well.

Moreover, the help texts and descriptions are (hopefully) simplified.


EDIT: superseded by:
- #953

mbolivar reviewed (commented):
Seems reasonable to me

pdgendt reviewed (approved):
Looks good!

carlescufi commented on src/west/app/project.py:
```diff
@@ -187,31 +218,39 @@ def do_add_parser(self, parser_adder):
+                            help=f'''manifest file to use. It is the relative
+                            path of the manifest file within the repository
+                            (remote or local). Defaults to {_WEST_YML}.''')
+        parser.add_argument('-t', '--topdir', dest='topdir',
```
We cannot get rid of (not even deprecate!) the positional argument, that would be a massive breaking change. I assume you implemented what @pdgendt suggested above, but personally I still think that `—topdir` should be only used with `-l`

carlescufi reviewed (approved):
Even though I'd prefer `--topdir` was only compatible with `-l`, this is not a blocking issue for me. Thanks!

marc-hb commented on src/west/app/project.py:
```diff
@@ -155,19 +156,49 @@ def __init__(self):
+1. Remote Manifest Repositories
+----------------------------
+If `-m` is given, west clones the manifest repository from the specified URL
+into the workspace. Config option manifest.path will point at this cloned
```
```suggestion
into a new workspace. Inside that workspace, the config option `manifest.path` will be the relative path to this git clone.
```

marc-hb commented on src/west/app/project.py:
```diff
@@ -155,19 +156,49 @@ def __init__(self):
 
-Warning: 'west init' renames and/or deletes temporary files inside the
+2. Local Manifest
+--------------
```
```suggestion
---------------
```

marc-hb commented on src/west/app/project.py:
```diff
@@ -192,9 +223,10 @@ def do_add_parser(self, parser_adder):
+  remote repository:
+  %(prog)s [-m URL] [--mr REVISION] [--mf FILE] [-o=GIT_CLONE_OPTION] [-t WORKSPACE_DIR] [directory]
+  local manifest:
+  %(prog)s -l [-t WORKSPACE_DIR] [--mf FILE] directory
```
```suggestion
  %(prog)s -l [-t WORKSPACE_DIR] [--mf FILE] [manifest_directory]
```

marc-hb commented on src/west/app/project.py:
```diff
@@ -155,19 +156,49 @@ def __init__(self):
+- If `--mr` is omitted, the repository's default branch (if available) is used.
 
-With neither, -m {MANIFEST_URL_DEFAULT} is assumed.
+Note: The west workspace will be created in the current working directory, if
```
```suggestion
Note: The west workspace will be the current working directory, if
```

I think this help text needs to entirely ban the expression "workspace in X/" because this expression is ambiguous: it can mean either `X/.west/` or `X/Y/.west`



This is even more important with this new `--topdir` option.

marc-hb reviewed (changes requested):
> This is particularly useful when using --local

I downloaded this PR and tested the other, --manifest-url option and it does not seem to work for me. I ran out of time and did not try --local yet. I didn't look at the code either.

Describing directory nesting clearly and accurately is **not** easy - it is basically a "specification" except it has to be user-friendly too. So I will look at the code only when that "specification" is complete and unambiguous.

This help looks like the two last ones can be both used at the same time:
```
west init [-m URL] [--mr REVISION] [--mf FILE] [-o=GIT_CLONE_OPTION] [-t WORKSPACE_DIR] [directory]
```

But it's not actually possible:
```
FATAL ERROR: --topdir cannot be combined with positional argument [directory]
```

When using either separately, each one seems to have the same effect. So, there does not seem to be any new feature in the remote case, which means it's still not possible to `west init` a manifest two levels down or more in the remote case? Is this an oversight? The help does not seem to say. I don't see why `--local` would be more flexible than `--manifest-url`. I would hope the `--manifest-url` and `--local` options work the same immediately after the git clone operation.

marc-hb commented:
> > This is particularly useful when using --local
> 
> I downloaded this PR and tested the other, --manifest-url option and it does not seem to work for me. I ran out of time and did not try --local yet. I didn't look at the code either.
> 
> Describing directory nesting clearly and accurately is **not** easy - it is basically a "specification" except it has to be user-friendly too. So I will look at the code only when that "specification" is complete and unambiguous (EDIT and of course: working for me).
> 
> This help looks like the two last ones can be both used at the same time:
> 
> ```
> west init [-m URL] [--mr REVISION] [--mf FILE] [-o=GIT_CLONE_OPTION] [-t WORKSPACE_DIR] [directory]
> ```
> 
> But it's not actually possible:
> 
> ```
> FATAL ERROR: --topdir cannot be combined with positional argument [directory]
> ```
> 
> When using either separately, each one seems to have the same effect. So, there does not seem to be any new feature in the remote case, which means it's still not possible to `west init` a manifest two levels down or more in the remote case? Is this an oversight? The help does not seem to say. I don't see why `--local` would be more flexible than `--manifest-url`. I would hope the `--manifest-url` and `--local` options work the same immediately after the git clone operation.

marc-hb commented on src/west/app/project.py:
```diff
@@ -155,19 +156,54 @@ def __init__(self):
+If `-m / --manifest-url` is given, west clones the manifest repository from the
+specified URL into a new workspace. Inside that workspace, the local config
+option `manifest.path` will be set to point at this cloned directory (relative
+path from the git clone to the workspace).
```
```suggestion
path to the git clone from the workspace).
```

marc-hb commented on src/west/app/project.py:
```diff
@@ -155,19 +156,61 @@ def __init__(self):
+in positional argument `[manifest_directory]` (it defaults to current working
+directory if no value is provided).
+Inside the new workspace, the local config option `manifest.path` is set to
+point at the directory containing the manifest (relative path to the workspace)
```
This is duplicated across 1. and 2. I would move it at the top, just before the "Arguments" section.

marc-hb commented on src/west/app/project.py:
```diff
@@ -155,19 +156,61 @@ def __init__(self):
+- the current working directory (if neither `--topdir` nor `[directory]` are
+  provided)
+
+Note:
```
```suggestion
```
A "Note" sounds like something not important to know. These seem important to me. And the whole section is not that long.

marc-hb commented on src/west/app/project.py:
```diff
@@ -155,19 +156,61 @@ def __init__(self):
+Inside the new workspace, the local config option `manifest.path` is set to
+point at the directory containing the manifest (relative path to the workspace)
+
+Note: The west workspace is created in the `manifest_directory`'s parent
```
```suggestion
The west workspace is created in the `manifest_directory`'s parent
```

marc-hb reviewed (dismissed):
I gave the new version a try and I think I found two confirmed bugs, see below. Reproducing and describing these bugs was very time-consuming so I ran out of time again and I didn't look at the code again :-(

I love the energy but please be much more thorough in testing edge cases and addressing review comments, otherwise the limited review bandwidth cannot keep up. It's a bit counter-intuitive but in this case "slowing down" means finishing sooner.

There seems to be a large quantity of test code being added which is great! However the "quality" of the test code matters even more than its quantity. I mean being imaginative and trying hard to "break" the production code by reaching as many edge cases as possible is much more important than being repetitive. Too much test repetition only makes running the test suite longer.


**Bug 1: spurious level in remote --topdir**

```
# west init  --topdir ws/   ws/manifest_dir/

$ tree -a -L 3 ws

ws
├── manifest_dir
│   └── zephyr      # <=== !!
│     ├── .git
│     
└── .west
    └── config
```

I don't think this extra "zephyr" level matches the current --help text. Either way, I don't think the extra level is the correct behavior because the extra level is inconsistent with how the --local behavior treat the positional argument.

Funny enough, this rejoins my other observation that writing "workspace/manifest created '**in**' X/" is ambiguous and a bad idea. Does "in" mean "X/.git" or "X/manifest/.git"? Unclear, avoid "in".


**Bug 2: re-initialization is not always caught**

I suspect the check that catches the re-initialization needs to be "upgraded" for the new --topdir


```
west init ws

cd ws
west init --local zephyr
FATAL ERROR: already initialized in ...                       # Good!

cd ..
west init --local --topdir ws/  ws/zephyr/
=== Initializing from existing manifest repository zephyr     # What?
--- Creating ws/.west and local configuration file
=== Initialized. Now run "west update" inside ws.
```

Interestingly enough, .west/config does not seem to be modified - unlike what the logs pretend. So there may be **more than one bug there**. I mean check why the logs are lying BEFORE fixing the check.

thorsten-klein commented:
> I suspect the check that catches the re-initialization needs to be "upgraded" for the new --topdir

@marc-hb I’ve tried with the code to fail if the topdir is already initialized (when using `--local`).
However, it seems that some tests (e.g. `test_update_some_with_imports`) rely on the current behavior and perform a re-initialization of some already initialized workspace.
What shall I do? 
It looks like this behavior might have been intentional — should I make it fail only when `args.topdir` is explicitly specified?

thorsten-klein commented:
> Has it? Please share one example of a west init --manifest-url ... command that can produce this
```
workspace
├── extra
│   └── level
│       └── manifest_clone
│           └── .git
└── .west
```

You requested that west is able to create such a layout. The command for this is now:
`west init  --topdir ws   ws/extra/level`

All projects will be cloned under `ws/extra/level`.
Is this correct behavior or what would be your expectation when both `args.topdir` and `directory` are provided?

marc-hb commented on src/west/app/project.py:
```diff
@@ -156,19 +157,60 @@ def __init__(self):
+    Defaults to `{_WEST_YML}` if not provided.
+
+-t / --topdir
+    Specifies the directory where west should create the workspace.
```
```suggestion
    Specifies the directory where west should create the workspace. It may already exist.
```

marc-hb commented on src/west/app/project.py:
```diff
@@ -156,19 +157,60 @@ def __init__(self):
+
+If both `-t / --topdir` and `directory` are provided, `-t / --topdir`
+specifies the workspace directory, while `directory` specifies the subfolder
+within the workspace where the remote repository is cloned during
```
```suggestion
within the workspace where the manifest repository is cloned during
```

marc-hb commented on src/west/app/project.py:
```diff
@@ -156,19 +157,60 @@ def __init__(self):
+If no `-m / --manifest-url` is provided, west uses Zephyr URL by default:
+  {MANIFEST_URL_DEFAULT}.
+
+The topdir (where `.west` is created) is determined as follows:
```
(It really sucks we have to keep such a complicated logic for backwards compatibility)

marc-hb replied:
Just found this, funny enough: https://interrupt.memfault.com/blog/practical_zephyr_west

> Notice that we no longer specify the current directory in the call to west init using “.”. In fact, the directory - optionally passed as the last argument to west init - is interpreted differently by West when using the --local flag or the -m arguments:
>
> With --local, the directory specifies the path to the local manifest repository.
Without the --local flag, the directory refers to the topdir and thus the folder in which to create the workspace (defaulting to the current working directory in this case).
> Awkward, but this approach is probably used due to legacy reasons. If no --local flag is used and no repository is specified using -m, West defaults to using the Zephyr repository.

marc-hb commented on src/west/app/project.py:
```diff
@@ -156,19 +157,60 @@ def __init__(self):
+2. Local Manifest
+-----------------
+If `-l / --local` is given, west initializes a workspace around an already
+existing `west.yml`. Its location can be specified in `manifest_directory`
```
```suggestion
existing `west.yml`. Its location can be specified in the `manifest_directory` positional argument
```

marc-hb commented on src/west/app/project.py:
```diff
@@ -156,19 +157,60 @@ def __init__(self):
+-----------------
+If `-l / --local` is given, west initializes a workspace around an already
+existing `west.yml`. Its location can be specified in `manifest_directory`
+(defaults to current working directory). Config option `manifest.path` is set
```
```suggestion
It defaults to current working directory.
```
No parenthesis. Is mentioning `manifest.path` really required at all in this help text? I think the west documentation should be enough.

marc-hb commented on src/west/app/project.py:
```diff
@@ -303,16 +369,28 @@ def local(self, args) -> Path:
+        if args.topdir:
+            topdir = Path(args.topdir).resolve()
+            try:
+                already = util.west_topdir(topdir, fall_back=False)
```
```suggestion
                already = util.west_topdir(manifest_dir, fall_back=False)
```
We want to maximize the chances of finding an existing workspace, not minimize them.

Also: move the `manifest_file.is_relative_to(topdir):` check first.

thorsten-klein replied:
I think this is not correct.
E.g. `--topdir workspace --local /path/to/my/manifest` should only fail if `topdir` is already a workspace, not if any parent of the manifest is already a workspace.

marc-hb reviewed (commented):
Partial review, I didn't get a chance to look at the test code yet, sorry. Later this week.


> I’ve tried with the code to fail if the topdir is already initialized (when using --local). However, it seems that some tests (e.g. test_update_some_with_imports) rely on the current behavior and perform a re-initialization of some already initialized workspace.
> What shall I do? It looks like this behavior might have been intentional, so I make it fail only when the new argument --topdir is used.
> 

I'm amazed that such "reinit" situations exist, is it still the case? If yes, can you elaborate and provide a simple reproduction? This looks like a serious bug. Does it predate this PR? Does it simply deletes the existing `.west/` directory? I can't believe there is code doing that.

> You requested that west is able to create such a layout. 

I only requested feature parity between --local and not --local. This is not "my" use case, this is the same use case than --local but when cloning.

> Is this correct behavior or what would be your expectation when both args.topdir and directory are provided?
> If they are mutually exclusive (as requested above), your requested feature will not be possible.

I don't remember requesting these they should be mutually exclusive, did I? I only remember experiencing that they were mutually exclusive in a previous version while the --help presented them as not mutually exclusive. So I requested to correct the help: https://github.com/zephyrproject-rtos/west/pull/854#discussion_r2434378346
Obviously, that's not relevant any more.


I find the current user interface confusing in the cloning case but:
- it seems consistent with --local
- it seems backwards-compatible
- I can't think of a better one at this point.

thorsten-klein commented on src/west/app/project.py:
```diff
@@ -405,6 +496,7 @@ def bootstrap(self, args) -> Path:
             # is PurePosixPath.
             manifest_path = PurePath(urlparse(manifest_url).path).name
 
+        manifest_path = str(Path(subdir) / manifest_path)
```
For setting the option the value have to be `str`: `TypeError: option values must be strings`
I have moved to conversion to the function call now 👍

marc-hb commented on src/west/app/project.py:
```diff
@@ -155,19 +156,61 @@ def __init__(self):
+--mf
+    The relative path to the manifest file within the manifest repository
+    (remote or local). Config option manifest.file will be set to this value.
+    Defaults to `{_WEST_YML}` if not provided.
```
Backquotes are useful in .rst docs but they look weird. Not important.

marc-hb commented on src/west/app/project.py:
```diff
@@ -156,19 +157,57 @@ def __init__(self):
+
+If both `-t / --topdir` and `directory` are provided, `-t / --topdir`
+specifies the workspace directory, while `directory` specifies the subfolder
+within the workspace where the repository is cloned during initialization
```
```suggestion
within the workspace where the manifest repository is cloned during initialization
```

marc-hb commented on src/west/app/project.py:
```diff
@@ -156,19 +157,57 @@ def __init__(self):
+
+The topdir (where `.west` is created) is determined as follows:
+- argument `-t / --topdir` if provided
+- otherwise: argument `directory` if provided
```
```suggestion
- otherwise: the positional `directory` argument if provided without --topdir. Deprecated: prefer --topdir.
```

marc-hb commented on src/west/app/project.py:
```diff
@@ -250,9 +310,15 @@ def do_add_parser(self, parser_adder):
+            metavar='directory',
+            help='''With --local: the path to a local directory which contains
+                    a west manifest (west.yml);
+                    Otherwise (if -m is used and no --topdir is provided):
```
Wait: `-m` is not relevant here, is it?
```suggestion
                    Otherwise, if no --topdir is provided:
```

marc-hb commented on src/west/app/project.py:
```diff
@@ -396,15 +481,17 @@ def bootstrap(self, args) -> Path:
-            manifest_path = PurePath(urlparse(manifest_url).path).name
+            mf_path = PurePath(urlparse(manifest_url).path).name
 
+        manifest_path: Path = Path(subdir) / mf_path
```
This looks like a misunderstanding... `manifest.path` comes either from the command line, or from `self.path` in the manifest file, but it's never a combination of both. If both are defined, the code should fail if they don't match.

If none is defined, keeping falling back on basename(url) like before.

marc-hb replied:
> Without prepending subdir the test fails...

Tests are only a means to an end, if they don't reflect the desired behavior then they need to change.

> As per definition

In the --local case, the positional argument becomes manifest.path. For consistency, I think we want the same in the cloning case. If that clashes with self.path then we should either error or warn, not sure which.

thorsten-klein replied:
I’m not sure we’re talking about the same thing. Let’s align.

When `west init --topdir subdir subdir/extra/level` is invoked, my expectation is that:

- The workspace is created under `subdir`, i.e. `subdir/.west` is created.
- The Zephyr repository is cloned to `<topdir>/extra/level/zephyr/.git`.
- The `manifest.path` configuration option points to `extra/level/zephyr`.
- The `manifest.file` configuration option is set to `west.yml`.

Does this match your expectations? Am I incorrect about any of these points, or missing something important?

marc-hb replied:
> The Zephyr repository is cloned to <topdir>/extra/level/zephyr/.git

No this does not match my expectations. For consistency with `west init -l topdir/extra/level/zephyr`, I'm expecting `west init --topdir topdir topdir/extra/level/zephyr` + an error or a warning when `self.path` is present and different from `extra/level/zephyr`. Long story short, all these arguments should always point at the manifest directory for consistency. Some arguments may be absolute, others like `self.path` must be relative but they should all point to the same directory in the end for simplicity. In case of a failure becaus of a conflict with `self.path`, the error message could require the users to 1. clone 2. west init in two steps.

Please prove me wrong and demonstrate that there is a more user-friendly design, but again it would have been more efficient to have this very discussion about the user interface and design first BEFORE writing the corresponding code, rather then "reserve-engineering" the code at code review time to discover that the UI design was not formalized.

This PR is 100% about nesting subdirectories, an area where "off by one" and similar errors are extremely common[*]. So it is critical to pay great attention to such nesting details. This is also why I repeatedly warned about phrasing like "thing gets created inside dir0" in the help text: does it mean "thing=dir0" or "dir0/thing"? No one knows. Less ambiguous language is required.

[*] examples:
- https://en.wikipedia.org/wiki/Tar_(computing)#Tarbomb
- https://unix.stackexchange.com/questions/402555/why-add-a-trailing-slash-after-an-rsync-destination
- Run `cp -r dir0/ dir1/` twice, starting with non-existent dir1. Results depend on the operating system!

carlescufi replied:
Let's try to figure out the actual disagreement here, step by step. First of all, let's use proper full commands. @marc-hb  and @thorsten-klein in the examples you are providing above with `west init —topdir` you are not including `-l` nor `-m`. Can we start by defining the full command-line arguments of the command that you @marc-hb think should work differently?

marc-hb reviewed (changes requested):
> manifest_path: Path = Path(subdir) / mf_path

This line does not make sense to me, see below.

thorsten-klein commented on src/west/app/project.py:
```diff
@@ -250,9 +306,15 @@ def do_add_parser(self, parser_adder):
+                    a west manifest (west.yml);
+                    Otherwise (if -m is used and no --topdir is provided):
+                    the location of the west workspace where .west directory
+                    will be created (<directory>/.west)
```
I would propose to do the deprecation in a separate PR.

thorsten-klein commented on src/west/app/project.py:
```diff
@@ -323,8 +395,21 @@ def local(self, args) -> Path:
+                    )
+                subdir = os.path.relpath(args.directory, topdir)
+        elif args.directory:
+            topdir = Path(abspath(args.directory))
```
I would propose to do the deprecation in a separate PR.

marc-hb commented:
> But I think in this particular case we are already too far ahead in the implementation and design to change the location of the discussions in my opinion

It's not really a "change" because all the `--topdir` user interface and high-level design discussion happened in #774 in the first place and the majority of it is still there. Even after this gets merged, someone wondering about the user interface would rather stick to #774 and not have to get lost in typos and other implementation details to find rationale(s).

BTW I'm writing "high-level design" as in "looking at the --help, not the code". This only about defining what the command-line parameters mean, it's not advanced maths! It's just faster to discuss and define this before writing and reviewing the code, that's all.

>  I don't think we are that far from agreeing to a solution here

Agreed, and I also love being proven wrong. I'm only upset about the time-consuming way to get there.

marc-hb reviewed (commented):
All the combinatorial explosion and complexity was clearly and vastly underestimated. I think the only way to get back on track is to follow the "top-down", more formal approach that should have been used in the first place:

1. Make an extensive list of all user input possibilities. NO CODE here.
2. Complete and unambiguous definition of all resulting behaviors that can immediately answer all questions like the ones above. NO CODE here.
3. "Tests first" approach with extensive coverage.
4. Code LAST (design first)

Steps 3. and 4. should be PAUSED until 1. and 2. are discussed and agreed: PLEASE DO NOT PUSH until they are. Prototyping is great. Reviewing prototypes while wrongly believing they are close to final has been crazy time-consuming, let's please stop?

marc-hb commented:
> [1] I did not look at these new force pushes yet. I bet even more will come.

There will be more revisions because the latest version https://github.com/zephyrproject-rtos/west/commit/cb9714f5d038299e8fd0129323c69af5d5ec8db9 fails on Windows:

https://github.com/zephyrproject-rtos/west/actions/runs/22708722014/job/65841352565?pr=854

```
       flags = _setup_test_init(manifest_repo, local_dir, topdir, directory, manifest_file)
        _, stderr = cmd_raises(['init'] + flags, expected_error)
>       assert expected_stderr in stderr
E       assert "FATAL ERROR: directory 'extra/dir' must be inside west topdir" in "FATAL ERROR: directory 'extra\\dir' must be inside west topdir 'D:\\a\\west\\west\\check out\\somedir'\n"
```

As a coincidence, I just had to work on similar portability issues and found a way to test Windows without Windows:
- #923

thorsten-klein commented:
> Can you give some reason(s) why you prefer this variant

I do not prefer this as well. Until now I thought that this was existing behavior. But it seems to be not. 
So I agree that it should be as you suggest 👍🏻

thorsten-klein committed:
add extensive tests for 'west init' and 'west init --topdir'

Added new argument for west init to specify the west topdir when a
workspace is inizialized.

marc-hb commented:
Fix #936 merged, @thorsten-klein please rebase. 

Also left TODO:

> - Once the behavior for absolute paths is finalized, there is the delicate question of relative paths.

- Using `anytree` in the test code (if needed)

I will investigate Windows-specific test failure if any left after rebase.

marc-hb commented on tests/test_project.py:
```diff
@@ -1876,6 +1871,261 @@ def test_init_local_with_manifest_filename(repos_tmpdir):
+
+    # initialize west workspace again
+    _, stderr = cmd_raises(['init', 'ws', '-m', zephyr], SystemExit)
+    assert "FATAL ERROR: already initialized" in stderr
```
De-duplicate using a constant like in the existing tests
```suggestion
    err_msg = "FATAL ERROR: already initialized"
    assert err_msg in stderr
```

marc-hb commented on tests/test_project.py:
```diff
@@ -1876,6 +1871,261 @@ def test_init_local_with_manifest_filename(repos_tmpdir):
+    zephyr = repos_tmpdir / 'repos' / 'zephyr'
+
+    # initial clone
+    cmd(['init', 'ws', '-m', zephyr])
```
Could use `--topdir` here too for a little bit more test coverage?

marc-hb commented on tests/test_project.py:
```diff
@@ -1876,6 +1871,261 @@ def test_init_local_with_manifest_filename(repos_tmpdir):
+    cmd(['init', 'ws', '-m', zephyr])
+
+    # initialize west workspace again
+    _, stderr = cmd_raises(['init', 'ws', '-m', zephyr], SystemExit)
```
Does this one add coverage compared to the existing `init_again` test?

thorsten-klein committed:
add extensive tests for 'west init'

multiple combinations of arguments are tested as they can be provided to
'west init'.

thorsten-klein committed:
support 'west init --topdir'

Added new argument for west init to specify the west topdir when a
workspace is inizialized.

thorsten-klein commented:
Apologies for the delay on this PR 👎🏻 This PR got out of control with all those (outdated) comments, commits and conversations.
I've updated it now and I hope I have tackled most concerns.

To keep things moving, I'd be happy if we can merge it as it is (if there are no bigger doubts).
Future comments we can tackle in a follow-up PR.

Alternatively, I'd propose a fresh start of this whole PR, so I will close this PR and submit a new one (so review process starts from scratch).

marc-hb commented:
Thanks @thorsten-klein for not giving up, much appreciated.

> Alternatively, I'd propose a fresh start of this whole PR, so I will close this PR and submit a new one (so review process starts from scratch).

I think this is probably the best next step at this stage. We can (and should) still refer to this older PR if needed, it's not going to disappear! But it will be easier for the new PR to focus on the (hopefully) very few important things left to discuss.

Also, please, please focus on a "top-down" approach in such a new PR, with a very heavy focus on all the _user interface_ details, corner cases and help text. It really think a "bottoms-up" approach was the main factor drowning this PR, where I kept trying to "reverse-engineer" the (sometimes unintentional) behavior from the code. Incredibly time-consuming.

Notably this (hopefully last) user interface question: https://github.com/zephyrproject-rtos/west/pull/854#issuecomment-4003176128

> To keep things moving, I'd be happy if we can merge it as it is (if there are no bigger doubts). Future comments we can tackle in a follow-up PR.

This approach works only when we are 100% sure that such follow-up PRs are clearly scoped, small and self-contained/isolated. Not when we are still discussing _user interface_ details. So not in this particular case I'm afraid.

thorsten-klein commented:
> Also, please, please focus on a "top-down" approach in such a new PR

@marc-hb I'll open a new PR with the current commits as they are now. Ok?

marc-hb commented:
> @marc-hb I'll open a new PR with the current commits as they are now. Ok?

Thanks, that should be a good starting point.

EDIT:
- #953
